internal/controller/persistentvolumeclaim_controller.go

```diff
@@ -66,19 +57,88 @@ func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctr
 		}, nil
 	}
 
-	// Skip if the PVC is not deleted or PVC does not have TopoLVM's finalizer.
-	if pvc.DeletionTimestamp == nil {
+	// Skip if PVC does not have TopoLVM's finalizer which added by the Pod mutating webhook.
+	if !controllerutil.ContainsFinalizer(pvc, topolvm.PVCFinalizer) {
 		return ctrl.Result{}, nil
 	}
 
-	if !controllerutil.ContainsFinalizer(pvc, topolvm.PVCFinalizer) {
-		return ctrl.Result{}, nil
+	if pvc.DeletionTimestamp == nil {
+		err := r.notifyKubeletToResizeFS(ctx, pvc)
+		return ctrl.Result{}, err
+	}
+
+	result, err := r.deletePodsUsingDeletingPVC(ctx, pvc)
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	updated := controllerutil.RemoveFinalizer(pvc, topolvm.PVCFinalizer)
+	if updated {
+		if err := r.client.Update(ctx, pvc); err != nil {
+			log.Error(err, "failed to remove finalizer", "name", pvc.Name)
+			return ctrl.Result{}, err
+		}
 	}
 
+	return ctrl.Result{}, nil
+}
+
+func (r *PersistentVolumeClaimReconciler) notifyKubeletToResizeFS(ctx context.Context, pvc *corev1.PersistentVolumeClaim) error {
+	// kubelet periodically checks pods to resize fs but it may take a long time.
+	// To speed up the process, mutate pods by adding an annotation.
+	// This process is only done when the PVC has `FileSystemResizePending` condition
+	// and pods are not mutated yet.
+	log := crlog.FromContext(ctx)
+	var condTransitionAt time.Time
+	for _, cond := range pvc.Status.Conditions {
+		if cond.Type == corev1.PersistentVolumeClaimFileSystemResizePending && cond.Status == corev1.ConditionTrue {
+			condTransitionAt = cond.LastTransitionTime.Time
+			goto needResizing
+		}
+	}
+	return nil
+
+needResizing:
+	pods, err := r.getPodsByPVC(ctx, pvc)
+	if err != nil {
+		log.Error(err, "unable to fetch PodList for a PVC")
+		return err
+	}
+	for _, pod := range pods {
+		if pod.Annotations == nil {
+			pod.Annotations = map[string]string{}
+		}
+
+		lastresizeFSRequestedAt, err := time.Parse(time.RFC3339Nano, pod.Annotations[topolvm.LastResizeFSRequestedAtKey])
+		if err != nil && lastresizeFSRequestedAt.After(condTransitionAt) {
+			continue
+		}
```

Comment:
The case where `err != nil` is true and `lastresizeFSRequestedAt.After(condTransitionAt)` is false is ignored.
I think we need an error handling.

Reply:
In this case, someone or a bug may set a wrong value to the annotation. Therefore, we can not trust the value, simply add a new one. This is the same logic as the happy path.